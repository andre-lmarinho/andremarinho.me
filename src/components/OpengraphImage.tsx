import { readFile } from 'node:fs/promises';
import type { Font } from 'next/dist/compiled/@vercel/og/satori';

type OGProps = {
  description: string;
  title: string;
  url?: string;
};

const assetUrl = (path: string) => new URL(`../../public${path}`, import.meta.url);

const fontSpecs: Array<[Font['name'], NonNullable<Font['weight']>, string]> = [
  ['Inter', 500, '/fonts/opengraph/Inter-Medium.woff'],
  ['Inter', 800, '/fonts/opengraph/Inter-ExtraBold.woff'],
  ['Roboto Mono', 400, '/fonts/opengraph/RobotoMono-Regular.woff'],
];

const fontsPromise: Promise<Font[]> = Promise.all(
  fontSpecs.map(async ([name, weight, path]) => ({
    data: await readFile(assetUrl(path)),
    name,
    style: 'normal',
    weight,
  }))
);

const profileImageSrcPromise = readFile(assetUrl('/images/Me.jpeg')).then(
  (buffer) => `data:image/jpeg;base64,${buffer.toString('base64')}`
);

type OpengraphImageProps = OGProps & { profileImageSrc: string };

function OpengraphImage({ title, description, url, profileImageSrc }: OpengraphImageProps) {
  return (
    <div
      style={{
        backgroundColor: 'white',
        width: '100%',
        height: '100%',
        display: 'flex',
        position: 'relative',
        fontFamily: 'Inter',
        backgroundImage: 'radial-gradient(at 10px 10px, #d4d4d4, #d4d4d4 5%, white 5%)',
        backgroundSize: '20px 20px',
      }}
    >
      <img
        src={profileImageSrc}
        width="115"
        height="115"
        style={{
          borderRadius: '100%',
          position: 'absolute',
          left: 60,
          top: 60,
        }}
      />
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          gap: '0.5rem',
          position: 'absolute',
          bottom: 140,
          left: 60,
          width: '65%',
        }}
      >
        <span
          style={{
            fontSize: '5.25rem',
            fontWeight: 700,
            letterSpacing: '-2px',
            lineHeight: '1',
          }}
        >
          {title}
        </span>
        <span
          style={{
            fontSize: '2.5rem',
            fontWeight: 400,
          }}
        >
          {description}
        </span>
      </div>
      <div
        style={{
          position: 'absolute',
          bottom: 60,
          left: 60,
          borderRadius: 100,
          fontFamily: 'RobotoMono',
          fontSize: '1.75rem',
          lineHeight: 1,
          fontWeight: 600,
        }}
      >
        {url ? `andremarinho.me/${url}` : 'andremarinho.me'}
      </div>
    </div>
  );
}

export async function buildOg(props: OGProps) {
  const [fonts, profileImageSrc] = await Promise.all([fontsPromise, profileImageSrcPromise]);
  const el = <OpengraphImage {...props} profileImageSrc={profileImageSrc} />;
  return [el, { fonts }] as const;
}
