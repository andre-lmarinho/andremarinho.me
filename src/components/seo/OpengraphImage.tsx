import { siteName, siteUrl } from '@/config/metadata';
import {
  interExtraBoldFontBase64,
  interMediumFontBase64,
  interRegularFontBase64,
} from '@/components/seo/ogFonts';

type FontWeight = 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900;

type FontSource = {
  name: string;
  data: ArrayBuffer;
  style?: 'normal' | 'italic';
  weight?: FontWeight;
};

type Props = {
  description: string;
  title: string;
  url?: string;
};

const decodeFont = (base64: string): ArrayBuffer => {
  if (typeof Buffer !== 'undefined') {
    const buffer = Buffer.from(base64, 'base64');
    return buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
  }

  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);

  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }

  return bytes.buffer;
};

export const getFonts = (): FontSource[] => {
  return [
    {
      name: 'Inter',
      data: decodeFont(interRegularFontBase64),
      style: 'normal',
      weight: 400,
    },
    {
      name: 'Inter',
      data: decodeFont(interMediumFontBase64),
      style: 'normal',
      weight: 600,
    },
    {
      name: 'Inter',
      data: decodeFont(interExtraBoldFontBase64),
      style: 'normal',
      weight: 700,
    },
  ];
};

const domain = new URL(siteUrl).host;

const OpengraphImage = ({ description, title, url }: Props) => {
  const displayUrl = url
    ? `${domain}${url.startsWith('/') ? url : `/${url}`}`
    : domain;

  return (
    <div
      style={{
        width: '100%',
        height: '100%',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-between',
        backgroundColor: '#0b1120',
        backgroundImage:
          'radial-gradient(circle at 20% 20%, rgba(56,189,248,0.15), transparent 55%), radial-gradient(circle at 80% 0%, rgba(129,140,248,0.15), transparent 40%)',
        padding: '72px',
        color: '#e2e8f0',
        fontFamily: 'Inter',
        position: 'relative',
      }}
    >
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          gap: '24px',
        }}
      >
        <div
          style={{
            width: 128,
            height: 128,
            borderRadius: '50%',
            border: '6px solid rgba(148,163,184,0.25)',
            boxShadow: '0 24px 60px rgba(15,23,42,0.45)',
            background:
              'linear-gradient(135deg, rgba(56,189,248,0.8), rgba(129,140,248,0.8))',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            color: '#0b1120',
            fontWeight: 700,
            fontSize: '2rem',
          }}
        >
          AM
        </div>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            gap: '6px',
          }}
        >
          <span
            style={{
              fontSize: '3rem',
              fontWeight: 700,
              letterSpacing: '-1px',
            }}
          >
            {siteName}
          </span>
          <span
            style={{
              fontSize: '1.75rem',
              color: 'rgba(148,163,184,0.9)',
              fontFamily: 'Inter',
              letterSpacing: '0.05em',
              textTransform: 'uppercase',
            }}
          >
            Front-End Developer
          </span>
        </div>
      </div>

      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          gap: '32px',
          maxWidth: '720px',
        }}
      >
        <span
          style={{
            fontSize: '5.5rem',
            fontWeight: 700,
            lineHeight: 1,
            letterSpacing: '-3px',
          }}
        >
          {title}
        </span>
        <span
          style={{
            fontSize: '2.4rem',
            color: 'rgba(226,232,240,0.95)',
            lineHeight: 1.3,
            fontWeight: 400,
          }}
        >
          {description}
        </span>
      </div>

      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'space-between',
          fontSize: '1.8rem',
          fontFamily: 'Inter',
          fontWeight: 400,
          color: 'rgba(148,163,184,0.9)',
        }}
      >
        <span>{displayUrl}</span>
        <span
          style={{
            padding: '8px 20px',
            borderRadius: 999,
            border: '1px solid rgba(148,163,184,0.35)',
            fontSize: '1.5rem',
            letterSpacing: '0.08em',
          }}
        >
          andremarinho.me
        </span>
      </div>
    </div>
  );
};

export default OpengraphImage;
