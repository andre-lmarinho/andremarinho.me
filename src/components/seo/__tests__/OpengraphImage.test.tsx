import { render, screen } from '@testing-library/react';

import OpengraphImage from '@/components/seo/OpengraphImage';

describe('OpengraphImage', () => {
  it('matches the expected structure for the default card', () => {
    const { container } = render(
      <OpengraphImage
        title="Andre Marinho - Front-End Developer"
        description="I am a Front-End Developer based in Salvador. I create digital experiences that connect design, strategy and business growth."
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the scoped url when provided', () => {
    render(
      <OpengraphImage
        title="About me"
        description="Learn about Andre"
        url="about"
      />,
    );

    expect(screen.getByText('andremarinho.me/about')).toBeInTheDocument();
  });
});
