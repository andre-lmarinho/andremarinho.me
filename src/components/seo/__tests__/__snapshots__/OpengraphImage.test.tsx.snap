// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OpengraphImage matches the expected structure for the default card 1`] = `
<div
  style="width: 100%; height: 100%; display: flex; flex-direction: column; justify-content: space-between; background-color: rgb(11, 17, 32); background-image: radial-gradient(circle at 20% 20%, rgba(56,189,248,0.15), transparent 55%), radial-gradient(circle at 80% 0%, rgba(129,140,248,0.15), transparent 40%); padding: 72px; color: rgb(226, 232, 240); font-family: Inter; position: relative;"
>
  <div
    style="display: flex; align-items: center; gap: 24px;"
  >
    <div
      style="width: 128px; height: 128px; border-radius: 50%; border: 6px solid rgba(148, 163, 184, 0.25); box-shadow: 0 24px 60px rgba(15,23,42,0.45); background: linear-gradient(135deg, rgba(56,189,248,0.8), rgba(129,140,248,0.8)); display: flex; align-items: center; justify-content: center; color: rgb(11, 17, 32); font-weight: 700; font-size: 2rem;"
    >
      AM
    </div>
    <div
      style="display: flex; flex-direction: column; gap: 6px;"
    >
      <span
        style="font-size: 3rem; font-weight: 700; letter-spacing: -1px;"
      >
        Andre Marinho
      </span>
      <span
        style="font-size: 1.75rem; color: rgba(148, 163, 184, 0.9); font-family: Inter; letter-spacing: 0.05em; text-transform: uppercase;"
      >
        Front-End Developer
      </span>
    </div>
  </div>
  <div
    style="display: flex; flex-direction: column; gap: 32px; max-width: 720px;"
  >
    <span
      style="font-size: 5.5rem; font-weight: 700; line-height: 1; letter-spacing: -3px;"
    >
      Andre Marinho - Front-End Developer
    </span>
    <span
      style="font-size: 2.4rem; color: rgba(226, 232, 240, 0.95); line-height: 1.3; font-weight: 400;"
    >
      I am a Front-End Developer based in Salvador. I create digital experiences that connect design, strategy and business growth.
    </span>
  </div>
  <div
    style="display: flex; align-items: center; justify-content: space-between; font-size: 1.8rem; font-family: Inter; font-weight: 400; color: rgba(148, 163, 184, 0.9);"
  >
    <span>
      andremarinho.me
    </span>
    <span
      style="padding: 8px 20px; border-radius: 999px; border: 1px solid rgba(148, 163, 184, 0.35); font-size: 1.5rem; letter-spacing: 0.08em;"
    >
      andremarinho.me
    </span>
  </div>
</div>
`;
